mypyc/ir/rtypes.py

```diff
@@ -436,6 +446,131 @@ def deserialize(cls, data: JsonDict, ctx: 'DeserMaps') -> 'RTuple':
+    else:
+        assert False, "invalid rtype for computing size"
+
+
+def compute_aligned_offsets_and_size(types: List[RType]) -> Tuple[List[int], int]:
+    """Compute offsets and total size of a list of types after alignment
+
+    Note that the types argument are types of values that are stored
+    sequentially with platform default alignment.
+    """
+    unaligned_sizes = [compute_rtype_size(typ) for typ in types]
+    alignments = [compute_rtype_alignment(typ) for typ in types]
+
+    current_offset = 0
+    offsets = []
+    final_size = 0
+    for i in range(len(unaligned_sizes)):
+        offsets.append(current_offset)
+        if i + 1 < len(unaligned_sizes):
+            cur_size = unaligned_sizes[i]
+            current_offset += cur_size
+            next_alignment = alignments[i + 1]
+            # compute aligned offset,
+            # check https://en.wikipedia.org/wiki/Data_structure_alignment for more infomation
+            current_offset = (current_offset + (next_alignment - 1)) & -next_alignment
+        else:
+            struct_alignment = max(alignments)
+            final_size = current_offset + unaligned_sizes[i]
+            final_size = (final_size + (struct_alignment - 1)) & -struct_alignment
+    return offsets, final_size
+
+
+class StructInfo:
+    """Struct-like type Infomation
+
+    StructInfo should work with registry to ensure constraints like the unique naming
+    constraint for struct type
+    """
+    def __init__(self,
+                 name: str,
+                 names: List[str],
+                 types: List[RType]) -> None:
+        self.name = name
+        self.names = names
+        self.types = types
+        # generate dummy names
+        if len(self.names) < len(self.types):
+            for i in range(len(self.types) - len(self.names)):
+                self.names.append('_item' + str(i))
+        self.offsets, self.size = compute_aligned_offsets_and_size(types)
+
+
+class RStruct(RType):
+    """Represent CPython structs"""
+    def __init__(self,
+                 info: StructInfo) -> None:
+        self.name = info.name
+        self.names = info.names
+        self.types = info.types
+        self.offsets = info.offsets
+        self.size = info.size
+        self._ctype = info.name
+
+    def accept(self, visitor: 'RTypeVisitor[T]') -> T:
+        return visitor.visit_rstruct(self)
+
+    def __str__(self) -> str:
+        # if not tuple(unamed structs)
+        return '%s{%s}' % (self.name, ', '.join(name + ":" + str(typ)
+                                                for name, typ in zip(self.names, self.types)))
+
+    def __repr__(self) -> str:
+        return '<RStruct %s{%s}>' % (self.name, ', '.join(name + ":" + repr(typ) for name, typ
+                                                          in zip(self.names, self.types)))
+
+    def __eq__(self, other: object) -> bool:
+        return (isinstance(other, RStruct) and self.name == other.name
```

Comment:
It's sufficient to compare the identity of the `info` objects if you implement the change I'm suggesting above.